### PR TITLE
Feature/improve state unitary retrieval for aer and qulacs

### PR DESCRIPTION
**Pull request by yao-cqc**
``QulacsBackend``, ``AerStateBackend`` and ``AerUnitaryBackend`` now take care of implicit qubit permutation and wire endianness before running circuits. The result retrieval speed is now instant given default permutation and endianness.

### CODE REVIEW

File: modules/pytket-qiskit/pytket/extensions/qiskit/backends/aer.py [hunk 1]
```diff
@@ -159,13 +159,23 @@ def process_circuits(
...
             elif self._backend_info.device_name == "aer_simulator_unitary":
+                qcs = [
+                    tk_to_qiskit(tkc, reverse_index=True, replace_implicit_swaps=True)
+                    for tkc in batch
+                ]
                 for qc in qcs:
                     qc.save_unitary()
+            else:
+                qcs = [tk_to_qiskit(tkc) for tkc in batch]
+
```
Comment (ss2165) [inline modules/pytket-qiskit/pytket/extensions/qiskit/backends/aer.py:178]:
```suggestion
        if self.supports_state:
            reverse_index, replace_implicit_swaps = (True, True)
            for qc in qcs:
                qc.save_state()

        elif self.supports_unitary:
            reverse_index, replace_implicit_swaps = (True, True)
            for qc in qcs:
                qc.save_unitary()

        qcs = [
            tk_to_qiskit(
                tkc, reverse_index=reverse_index, replace_implicit_swaps=replace_implicit_swaps
            )
            for tkc in batch
        ]


```

**Review by ss2165 — CHANGES_REQUESTED** [review]
Looks good! I think it would be good to have a test of the "replace_implicit_swaps" method in at least one of the backends as a sanity check. You could refactor it out in to a method that returns a list of pairs of integers and test that independently. Then use those integers to apply swaps.

**Comment by yao-cqc** [comment]
As for the code duplication, I think it might be worth to add a pass to convert implicit swaps to actual swaps in tket as well as a method to convert the endianness of the circuit (similar to Qiskit's ``reverse_bits``).

File: modules/pytket-qiskit/pytket/extensions/qiskit/backends/aer.py [hunk 2]
```diff
@@ -158,24 +158,19 @@ def process_circuits(
...
 
+        for (n_shots, batch), indices in zip(circuit_batches, batch_order):
+            qcs = []
+            for tkc in batch:
+                qc = tk_to_qiskit(tkc, reverse_index, replace_implicit_swaps)
+                if self.supports_state:
+                    qc.save_state()
+                elif self.supports_unitary:
+                    qc.save_unitary()
+                qcs.append(qc)
```
Comment (ss2165) [inline modules/pytket-qiskit/pytket/extensions/qiskit/backends/aer.py:173]: nice 👍

**Review by ss2165 — APPROVED** [review]

**Pull request merged**